File: app/src/main/java/peter/util/searcher/adapter/MultiWindowAdapter.java
package peter.util.searcher.adapter;

import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;

import peter.util.searcher.R;
import peter.util.searcher.activity.MainActivity;
import peter.util.searcher.tab.TabGroup;

public class MultiWindowAdapter extends BaseAdapter {

    private ArrayList<TabGroup> mList;

    private MainActivity mainActivity;

    public void update(MainActivity activity) {
        mainActivity = activity;
        mList = mainActivity.getManager().getList();
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        if (mList != null) {
            return mList.size();
        }
        return 0;
    }

    @Override
    public TabGroup getItem(int position) {
        if (mList != null) {
            return mList.get(position);
        }
        return null;
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {

        Holder holder = null;
        if (convertView == null) {
            convertView = mainActivity.getLayoutInflater().inflate(R.layout.multiwindow_item, parent, false);
            holder = new Holder();
            holder.close = (ImageView) convertView.findViewById(R.id.close_tab);
            holder.title = (TextView) convertView.findViewById(R.id.title);
            holder.icon = (ImageView) convertView.findViewById(R.id.icon);
            convertView.setTag(holder);
        }

        if (holder == null) {
            holder = (Holder) convertView.getTag();
        }

        TabGroup tabGroup = getItem(position);
        if (tabGroup != null) {
            if (mainActivity.getManager().getCurrentTabGroup() == tabGroup) {
                convertView.setActivated(true);
            } else {
                convertView.setActivated(false);
            }
            holder.title.setText(tabGroup.getCurrentTab().getTitle());
            Drawable icon = tabGroup.getCurrentTab().getIconDrawable();
            if (icon != null) {
                holder.icon.setBackground(icon);
            } else {
                holder.icon.setBackgroundResource(R.drawable.web_site_icon);
            }
            holder.close.setOnClickListener(mainActivity);
            holder.close.setTag(tabGroup);
            convertView.setOnClickListener(mainActivity);
            convertView.setTag(R.id.multi_window_item_tag, tabGroup);
        }
        return convertView;
    }

    private static class Holder {
        ImageView icon;
        TextView title;
        ImageView close;
    }

}

File: app/src/main/java/peter/util/searcher/adapter/MultiWindowAdapter.kt
package peter.util.searcher.adapter

import android.util.SparseArray
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView

import java.util.ArrayList

import peter.util.searcher.R
import peter.util.searcher.activity.MainActivity
import peter.util.searcher.tab.TabGroup

class MultiWindowAdapter : BaseAdapter() {

    private var mList: ArrayList<TabGroup> = ArrayList()
    private var mainActivity: MainActivity? = null

    fun update(activity: MainActivity) {
        mainActivity = activity
        mList = mainActivity!!.tabManager!!.list
        notifyDataSetChanged()
    }

    override fun getCount(): Int =  mList.size

    override fun getItem(position: Int): TabGroup? = mList[position]

    override fun getItemId(position: Int): Long = position.toLong()

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var content = convertView
        if (content == null) {
            content = mainActivity!!.layoutInflater.inflate(R.layout.multiwindow_item, parent, false)
        }

        val tabGroup = getItem(position)

        content!!.isActivated = mainActivity!!.tabManager!!.currentTabGroup() === tabGroup

        val title: TextView = content.findViewOften(R.id.title)
        title.text = tabGroup?.currentTab()?.getTitle()
        val icon: ImageView = content.findViewOften(R.id.icon)
        if (tabGroup?.currentTab()?.iconDrawable != null) {
            icon.background = tabGroup.currentTab()!!.iconDrawable
        } else {
            icon.setBackgroundResource(R.drawable.ic_website)
        }
        val close: ImageView = content.findViewOften(R.id.close_tab)
        close.tag = tabGroup
        close.setOnClickListener(mainActivity)
        content.setTag(R.id.multi_window_item_tag, tabGroup)
        return content
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T : View> View.findViewOften(viewId: Int): T {
        val viewHolder: SparseArray<View> = tag as? SparseArray<View> ?: SparseArray()
        tag = viewHolder
        var childView: View? = viewHolder.get(viewId)
        if (null == childView) {
            childView = findViewById(viewId)
            viewHolder.put(viewId, childView)
        }
        return childView as T
    }


}